Log and manage power controls of the charging pad

itslinotlie opened the pull request:
Current status:
- Able to read and respond to ```MSG_ACTUATOR_CMD``` CAN messages (toggling the 5V power)
- Able to log necessary info via ```MSG_SENSOR_ANALOG``` CAN messages (values are inline with measured voltage values)

Left to do:
- test the charging capabilities with a lipo (?)
    - this'll also test the ```SENSOR_BUS_CURR```, ```SENSOR_BATT_CURR```, and ```SENSOR_CHARGE_CURR``` values (?)
    
 Testing is planned for the weekend of March 25-26 and until then, this branch will remain dormant

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/waterloo-rocketry/cansw_charging/1)
<!-- Reviewable:end -->

itslinotlie committed:
include necessary mcc files

contents copied from cansw_actuator repo

itslinotlie committed:
Customize boiler plate

also added comments to map out what needs to be done

itslinotlie committed:
Add voltage health can messages

itslinotlie committed:
forgot multiplier for VSENSE

itslinotlie committed:
im done for tonight

need to figure out which adcc_channels I need and also what constant to scael the getSingleConversion value by

itslinotlie committed:
I found the adcc_channel values for my pins (:

itslinotlie committed:
talked to jack, no need to scale

this math is soooo clean. so basically, if your voltage reference is 4.096 and you have a 12 bit adc message, that comes to 0.001 mV and since we want volts thats just 1V so really, you don't have to scale anything and you get your voltage straight from the proper channels. waohhhhh

itslinotlie committed:
add remaining pin initialized values

itslinotlie committed:
thanks to joe, i think this is 'done'

need to test the default values for 5V_line and BATT_on to see if its a 0 or 1, and then also create (?) sensor IDs for the 5 and 13volt lines

itslinotlie committed:
add vscode to gitignore

QuantumManiac commented:
@itslinotlie Repo was private so features were restricted. Changed repo visibility to public and converted this PR into a draft.

itslinotlie committed:
change nplab project setting name to cansw_charging

itslinotlie committed:
configure mplab nbproject settings

mcc_genereated_files are now linked and I fixed up the file structure

itslinotlie committed:
code should be good now

probably need to create a new board sensor id for the 5v and 13v current lines?

itslinotlie committed:
Fixed voltage logging values

tldr; had to slow down the clock to sample the analog-to-digital-converter properly, only spent half a day (:

itslinotlie committed:
I think i mixed up my tris values

tris=1 is input, aka you read in while tris=0 is output, so you write the output of the pin

itslinotlie committed:
Board accurately responds to CAN messages

MSG_ACTUATOR_CMD ACTUATOR OPEN/CLOSED turns the 5V line on or off respectivelynow

itslinotlie committed:
Update can_msg sensor ids

need to create sensor_charge curr and volt in canlib for the sensor_id

itslinotlie committed:
Fix charging current leaking

basically I had the battery charger value set incorrectly (forgot to test)

itslinotlie committed:
Cleanup comments

itslinotlie committed:
turns out the sketchy adc channel stays

jack mentioned that creating the adcc channel enums in the adcc.h file was a bit sketchy, but since there are functions that reference the enum type, it needs to be defined there as well

itslinotlie committed:
rename variables to be clearer

itslinotlie committed:
Yoink cansw_arming low_pass_filter

Just need to verify this in person (can toggle low_pass_response to control the weightings and MAX_SENSOR_LOOP_TIME_DIFF_ms for how frequent to modify said average

itslinotlie committed:
Tested low pass filter

there was also a mistake with the current scalar factor, was off by a factor of 2

Commit:
Update code

canlib changed, board respin happened, fixed some issues where I was multipling instead of dividing (need to confirm by testing it in person)

Commit:
address bugs

Commit:
Update canlib commit id

BluCodeGH committed:
WDR fixes

BluCodeGH reviewed (approved):
LGTM!

Commit:
Remove TODOs

if it aint broken, why fix it
